### PR TITLE
Validate xpath

### PR DESCRIPTION
**Pull request by dannyroberts**
I'll put this on staging as a proof of concept for @amsagoff to play with.

Good xpath

screen shot 2015-12-05 at 1 32 04 pm

Bad xpath

screen shot 2015-12-05 at 1 32 23 pm

### CODE REVIEW

**Comment by amsagoff** [comment]
Oh my!!! One step closer to removing the feature preview. Will check this
out asap

**Comment by dannyroberts** [comment]
Great :) It's on staging for both Module Filter and the Form Display Condition

**Comment by benrudolph** [comment]
this is awesome! :candy:

File: corehq/apps/style/templates/style/bootstrap2/ko/value-or-none-ui.html [hunk 1]
```diff
@@ -43,18 +43,21 @@
         </div>
         <div data-bind="visible: allowed()">
             <input type="text" data-bind="
-                    value: inputValue,
+                    xpathValidator: inputValue,
```
Comment (benrudolph) [inline corehq/apps/style/templates/style/bootstrap2/ko/value-or-none-ui.html]: this is pretty neat

File: bower.json [hunk 2]
```diff
@@ -21,6 +21,7 @@
     "knockout": "3.1.0",
     "select2-3.4.5-legacy": "select2#3.4.5",
     "less": "1.7.3",
+    "xpath": "dimagi/js-xpath",
```
Comment (benrudolph) [inline bower.json]: :+1:

**Comment by snopoke** [comment]
:candy:

**Comment by dannyroberts** [comment]
I just played around with having the same validator run on the server side (which allows us to do build validation without relying on the client) and came up with this so far, which seems to work well. (From python we'd call it similar to how we call the validator jar.)

```javascript
# xpathValidator.node.js
var xpath = require('xpath');
require('models');
require('biginteger');
var SchemeNumber = require('schemeNumber').SchemeNumber;
var fs = require('fs');
var data = fs.readFileSync('/dev/stdin').toString();
try {
    xpath.parse(data);
} catch (e) {
    console.log(e.message);
    process.exit(1);
}
process.exit(0);
```
Which we can call from the command line like so:

```bash
$ echo "/data/node" | ( export NODE_PATH=./bower_components/xpath/lib:./bower_components/xpath/ && node xpathValidator.node.js )
$ echo $?
0
$ echo "a a" | ( export NODE_PATH=./bower_components/xpath/lib:./bower_components/xpath/ && node xpathValidator.node.js )
Lexical error on line 1. Unrecognized text.
a a
--^
$ echo $?
1
```

**Comment by czue** [comment]
interesting. if we're pushing the validator to server side could we just the validator jar? seems like using the one that runs on the phone would be an improvement to the js one, no?

**Comment by czue** [comment]
(forgot to mention but this is really cool!)

**Comment by benrudolph** [comment]
agree with cory, if we already have the same functionality in that jar. if not, might be slightly easier to put the xpath in the `package.json` so it also installs in `node_modules` and you don't have to worry about the path as much. (any node file should automatically search in `node_modules` for executables)

**Comment by dannyroberts** [comment]
@benrudolph I'm a total noob with js packaging and node and was just pulling whatever I could from examples on the internet to see if I could throw a working example, so thanks for the tip.

As for the validator jar, I can ask Clayton if that's exposed. The other concern I had was that we may want to extend this to validate not just strict xpath but our little extension of the language to include the #case / #form / #user shortcuts (which of course are then interpolated before packaging up for the mobile). We may end up dealing with that in a different way, but the fact that I could easily think of a concrete example made at least plausible that we would want (1) control over the exact formal language being validated and (2) complete parity with the validation used by the client, so doing a callout to the same exact code being run on the client seemed like a reasonable thing to do. It's also probably pretty easy to swap out later on if we change our minds.

**Comment by emord** [comment]
> The other concern I had was that we may want to extend this to validate not just strict xpath but our little extension of the language to include the #case / #form / #user shortcuts

@dannyroberts I'm actually probably already going to do this for case management in the form builder. Here's the branch I did that on https://github.com/dimagi/js-xpath/tree/hashtags

**Comment by dannyroberts** [comment]
@emord that's what I'm talkin' 'bout!

**Comment by czue** [comment]
:+1: makes sense to me and good with this plan

**Comment by orangejenny** [comment]
Made some minor comments on the recent commits. Is this good to go otherwise?

**Comment by emord** [comment]
Would be worth testing again to make sure it works with the new js-xpath library.

File: corehq/apps/app_manager/templates/app_manager/partials/xpathValidator.html [hunk 3]
```diff
@@ -0,0 +1,83 @@
+{% load hq_shared_tags %}
+<script src="{% static 'xpath/lib/biginteger.js' %}"></script>
+<script src="{% static 'xpath/lib/schemeNumber.js' %}"></script>
+<script src="{% static 'xpath/models.js"' %}></script>
```
Comment (emord) [inline corehq/apps/app_manager/templates/app_manager/partials/xpathValidator.html]: I don't think you need to include model.js anymore

**Comment by dannyroberts** [comment]
I need to revisit this PR, thanks for the reminder. I think there's not too much to go before it's worth merging, so I'll make some time this week to get on it

**Comment by orangejenny** [comment]
Cool - don't mean to rush you, I just noticed that this and app manager B3 are both on staging now and aren't really compatible, but if this is close to ready, I don't want to migrate it to B3 and then have it block on that PR.

**Comment by dannyroberts** [comment]
@orangejenny I think I fixed the merge of this + b3. It's on the https://github.com/dimagi/commcare-hq/commit/b3-app-manager+validate-xpath branch

**Comment by dannyroberts** [comment]
@amsagoff would you be ok with merging this before we have server-side validation? I think this is ready to go. (And the things it runs on are all advanced or flagged.)

**Comment by amsagoff** [comment]
Probably - are we still talking about hte original screenshot?

**Comment by dannyroberts** [comment]
Yeah no changes, still just an icon that shows up to the right of the text box

**Comment by orangejenny** [comment]
@dannyroberts sweet, thank you!

**Comment by amsagoff** [comment]
good to go in feature preview land - will need some UX love before wide release. we can do that when we do the server-side work?

**Comment by dannyroberts** [comment]
@amsagoff said this is ok to merge. I would encourage it!

**Comment by dannyroberts** [comment]
(Oops missed your comment.) Yeah we can add server-side validation (so error messages at the top of  form & module pages + error message on build) and then release

**Comment by orangejenny** [comment]
I love merging things that I haven't reviewed, but that other reasonable people have reviewed. Here we go!

**Comment by orangejenny** [comment]
oh wait nevermind, will wait on validation

**Comment by amsagoff** [comment]
To clarify: I'm good releasing as-is, under the feature preview. In order
to wide release, we'd need some UX improvements (which require doing the
validation server-side)

**Pull request merged**